braindecode/models/eegdeformer.py

```diff
@@ -0,0 +1,288 @@
+        x_dense = torch.cat(dense_feature, dim=-1)  # b, in_chan*depth
+        x = x.view(x.size(0), -1)  # b, in_chan*d_hidden_last_layer
+        emd = torch.cat(
+            (x, x_dense), dim=-1
+        )  # b, in_chan*(depth + d_hidden_last_layer)
+        return emd
+
+    @staticmethod
+    def get_info(x):
+        # x: b, k, l
+        x = torch.log(torch.mean(x.pow(2), dim=-1))
+        return x
+
+
+class EEGDeformer(EEGModuleMixin, nn.Module):
+    """Deformer model from [ding2024]_.
+
+    XXXX.
+
+    Parameters
+    ----------
+    temporal_kernel : int
+        Size of the temporal convolutional kernel.
+    num_kernel : int, optional
+        Number of kernels (filters) in the convolutional layer. Default is 64.
+    depth : int, optional
+        Depth of the transformer (number of layers). Default is 4.
+    heads : int, optional
+        Number of attention heads in the transformer. Default is 16.
+    mlp_dim : int, optional
+        Dimension of the hidden layer in the feedforward network. Default is 16.
+    dim_head : int, optional
+        Dimension of each attention head. Default is 16.
+    drop_prob : float, optional
+        Dropout rate. Default is 0.0
+
+    References
+    ----------
+    .. [ding2024] Ding, Y., Li, Y., Sun, H., Liu, R., Tong, C., & Guan, C. (2024).
+       EEG-Deformer: A Dense Convolutional Transformer for Brain-computer
+       Interfaces. arXiv preprint arXiv:2405.00719.
+    """
+
+    def __init__(
+        self,
+        temporal_kernel: int = 11,
+        num_kernel: int = 64,
+        n_layers: int = 4,
+        heads: int = 16,
+        mlp_dim: int = 16,
+        dim_head: int = 16,
+        drop_prob: float = 0.0,
+        activation: nn.Module = nn.ELU,
+        # braindecode
+        chs_info=None,
+        input_window_seconds=None,
+        sfreq=None,
+        n_chans=None,
+        n_times=None,
+        n_outputs=None,
+    ):
+        super().__init__(
+            n_chans=n_chans,
+            n_times=n_times,
+            n_outputs=n_outputs,
+            chs_info=chs_info,
+            input_window_seconds=input_window_seconds,
+            sfreq=sfreq,
+        )
+        del n_outputs, n_chans, chs_info, n_times, input_window_seconds, sfreq
+        # Variables
+        self.n_layers = n_layers
+        self.drop_prob = drop_prob
+        self.dim = int(0.5 * self.n_times)
+        self.hidden_size = int(num_kernel * int(self.dim * (0.5**n_layers))) + int(
+            num_kernel * n_layers
+        )
+
+        # Parameters
+        self.pos_embedding = nn.Parameter(torch.randn(1, num_kernel, self.dim))
+
+        # Layers
+        self.ensuredim = Rearrange("batch chan time -> batch 1 chan time")
+
+        self.cnn_encoder = nn.Sequential(
+            Conv2dWithConstraint(
+                in_channels=1,
+                out_channels=num_kernel,
+                kernel_size=(1, temporal_kernel),
+                padding=(0, int(0.5 * (temporal_kernel - 1))),
```

Comment:
Is this a bug or a feature? I would either use `temporal_kernel//2` or `"same"` but this would change the padding for even kernel_lengths.

Reply:
bug, I solved in the last commit